File: <100/074/074.cpp
#include<iostream>
#include<set>
#include<time.h>
using namespace std;
using ll = int64_t;

ll factorial(ll n){
  if(n <= 1) return 1;
  else return n * factorial(n-1);
}


ll digitFactorialSum(ll n){
  if(n == 0) return 0;
  ll r = factorial(n%10);
  return r + digitFactorialSum(n/10);
}


ll evalLoopSize(ll n){
  set<ll> S;
  while(S.count(n) == 0){
    S.insert(n);
    n = digitFactorialSum(n);
  }

  return S.size();
}


int main(){
  clock_t start = clock();

  ll upper_limit = 1e6;
  ll answer = 0;

  // naive brute-force without memorization; 10 sec to execute
  for(ll n = 1; n < upper_limit; ++n){
    ll loop_size = evalLoopSize(n);
    answer += loop_size == 60;
  }

  clock_t end = clock();
  const double time = static_cast<double>(end - start) / CLOCKS_PER_SEC;

  cout << answer << endl;
  cout << time << " sec" << endl;
  return 0;
}

File: <100/074/074.jl
function digit_factorial_sum(n :: Int) :: Int
  result = 0
  while 0 < n
    result += factorial(n%10)
    n ÷= 10
  end
  return result
end


function solve(upper_limit :: Int) :: Int
  count = 0
  for i ∈ 1:(upper_limit)
    s :: Set{Int} = Set([])
    while !in(i, s)
      push!(s, i)
      i = digit_factorial_sum(i)
    end
    if length(s) == 60
      count += 1
    end
  end

  return count
end


function main() :: Nothing
  upper_limit :: Int = 10^6 - 1
  upper_limit |> solve |> println
end


@time main()
